File: src/dataset/typescript/complex-numbers_exercism-wackerow/gemini/GeminiFlash_complex-numbers_v1.ts
export default class ComplexNumber {
  private readonly realValue: number;
  private readonly imagValue: number;

  constructor(real: number, imag: number) {
    this.realValue = real;
    this.imagValue = imag;
  }

  get real(): number {
    return this.realValue;
  }

  get imag(): number {
    return this.imagValue;
  }

  get abs(): number {
    const real = this.realValue;
    const imag = this.imagValue;
    return Math.hypot(real, imag);
  }

  get conj(): ComplexNumber {
    return new ComplexNumber(this.realValue, -this.imagValue);
  }

  get exp(): ComplexNumber {
    const expReal = Math.exp(this.realValue);
    const cosImag = Math.cos(this.imagValue);
    const sinImag = Math.sin(this.imagValue);

    return new ComplexNumber(
      expReal * cosImag,
      expReal * sinImag
    );
  }

  public add(that: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this.realValue + that.realValue,
      this.imagValue + that.imagValue
    );
  }

  public sub(that: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this.realValue - that.realValue,
      this.imagValue - that.imagValue
    );
  }

  public mul(that: ComplexNumber): ComplexNumber {
    const real1 = this.realValue;
    const imag1 = this.imagValue;
    const real2 = that.realValue;
    const imag2 = that.imagValue;

    return new ComplexNumber(
      real1 * real2 - imag1 * imag2,
      imag1 * real2 + real1 * imag2
    );
  }

  public div(that: ComplexNumber): ComplexNumber {
    const real1 = this.realValue;
    const imag1 = this.imagValue;
    const real2 = that.realValue;
    const imag2 = that.imagValue;

    const denominator = real2 * real2 + imag2 * imag2;

    return new ComplexNumber(
      (real1 * real2 + imag1 * imag2) / denominator,
      (imag1 * real2 - real1 * imag2) / denominator
    );
  }
}

File: src/dataset/cpp/complex-numbers_exercism-tqa236/gemini/GeminiFlash_complex_numbers_v1.cpp
#include "complex_numbers.h"

#include <cmath>

namespace complex_numbers {
Complex::Complex(double real_, double imag_) : real_(real_), imag_(imag_) {}
double Complex::real() const { return real_; }
double Complex::imag() const { return imag_; }
double Complex::abs() const { return std::hypot(real_, imag_); }
Complex Complex::operator+(Complex const &obj) const {
  return Complex(real_ + obj.real_, imag_ + obj.imag_);
}
Complex Complex::operator-(Complex const &obj) const {
  return Complex(real_ - obj.real_, imag_ - obj.imag_);
}
Complex Complex::operator*(Complex const &obj) const {
  const double real_part = real_ * obj.real_ - imag_ * obj.imag_;
  const double imag_part = imag_ * obj.real_ + real_ * obj.imag_;
  return Complex(real_part, imag_part);
}
Complex Complex::operator/(Complex const &obj) const {
  const double denominator = obj.real_ * obj.real_ + obj.imag_ * obj.imag_;
  const double real_part = (real_ * obj.real_ + imag_ * obj.imag_) / denominator;
  const double imag_part = (imag_ * obj.real_ - real_ * obj.imag_) / denominator;
  return Complex(real_part, imag_part);
}
Complex Complex::conj() const { return Complex(real_, -imag_); }
Complex Complex::exp() const {
  const double common_factor = std::exp(real_);
  return Complex(common_factor * std::cos(imag_), common_factor * std::sin(imag_));
}
}  // namespace complex_numbers